### DISCUSSION TITLE
Nested hydration

### DISCUSSION THREAD

**Discussion by ju5t** [livewire/livewire#4501] [1. Help]
This is our use case:

- We have a list of Models.
- We want to add, edit and delete them.
- We don't want the changes to be effective immediately. We want them to be saved when a user clicks save.
- We prefer access to the models, so we can access relationships.

It's intended to CRUD suppliers to an article. Yes, I just used CRUD as a verb :)

To retrieve the list of suppliers linked to the article we do:

```php
SupplierLink::query()
    ->where('article_id', $model->id)
    ->get();
```

There are two options here:

- Return `Illuminate\Database\Eloquent\Collection`
- Return `Illuminate\Support\Collection` by adding `->collect()` to the query.

Both work in terms of retrieving the suppliers and both work initially when requiring access to the model itself.

To support creating and updating we have the following logic:

```php
$this->supplierLinks = match ($supplierLink->id === null) {
    true => $this->supplierLinks->add($supplierLink),
    false => $this->supplierLinks->mapWithKeys(function ($item, $key) use ($supplierLink) {
        if ($item->id === $supplierLink->id) {
            return [$key => $supplierLink];
        }

        return [$key => $item];
    }),
};
```

When I'm using the first option and try to update the collection, it returns `Queueing collections with multiple model connections is not supported.`. I tried some of the workarounds, like setting the connection to MySQL when adding a new model, or initiating a new Eloquent Collection when creating the collection itself, but the errors didn't disappear.

edit: for some reason I am not seeing the error anymore mentioned above. It is having errors later on though, as the rehydration seems to forget all attributes and I'm left with an empty model.

Switching to the Support Collection breaks the requirement of having access to model relationships. Livewire only hydrates back to a collection, but doesn't support nesting. As such, we're left with a collection where each item is an array.

I don't know how to get around this, apart dropping one or more requirements. I'm hoping someone here knows how to solve this :)

**Comment by ju5t** [comment]
Let me clarify what made sure `Queueing collections with multiple model connections is not supported.` disappeared. Just in case someone finds this and wonders what was the fix.

For me it was changing:

```php
$this->supplierLinks->add($supplierLink)
```

into:

```php
$this->supplierLinks->push($supplierLink)
```

The full code for adding and updating items in a database collection is:

```php
$this->supplierLinks = match ($supplierLink->id === null) {
    true => $this->supplierLinks->push($supplierLink),
    false => $this->supplierLinks->mapWithKeys(function ($item, $key) use ($supplierLink) {
        if ($item->id === $supplierLink->id) {
            $item->fill($supplierLink->toArray());
        }

        return [$key => $item];
    }),
};
```

In this code `$supplierLink` is a `SupplierLink` model. 

I think there are only two issues at this point, both maybe related to the same part I probably don't understand. It has to do with the rehydration of models inside a database collection. If I update a model inside the collection and `@json` the collection in my browser, as soon as the XHR finishes, it shows the changed value. 

When I press save which calls a simple `dd` of the collection, it resets this value back to the original, coming from the database.

For new items I am seeing something similar. When I create a new item it shows in the `@json`. But this time, when I save the value, it resets to a new model and all values are gone.

I am missing something here but I don't know what it is. Is what I am trying not (yet) possible with Livewire? Is there an alternative for what I am trying to achieve?

**Comment by joshhanley** [comment]
@ju5t so the issue is actually an issue with Laravel's serialisation of collections. For some reason, it doesn't like it where there is an existing eloquent collection of models loaded from the DB and then a new model get's added that hasn't been persisted to the database yet. Hence the error `Queueing collections with multiple model connections is not supported.`.

I believe this is because it was never written to function that way, due to it being built for serialising models to put onto a queued job (in which case they would have already been persisted).

So while it's convenient to try and use the collections, I think if you need both create and update in the same collection, you'd be better off converting the data to an array, then create/update as required, then persist the changes back to the models on save.

In regards to the data resetting, make sure you have rules setup for all the properties that you want Livewire to keep track of, otherwise, LW will only restore the data from the db, but not apply the updates.

Hope this helps!

**Reply by ju5t** [reply]
@joshhanley thanks for your reply! The queueing error was fixed by switching from `$collection->add($item)` to `$collection->push($item)`. There are rules setup for all fields, but it would still reset the data.

I have changed everything to arrays now and with some magic dust here and there it all works. To be fair, updating the array is much simpler compared to collections.

```php
public function updateSupplierLink(array $data): void
{
    $key = $data['key'];
    $link = $data['link'];

    match (! is_null($key)) {
        true => Arr::set($this->supplierLinks, $key, $link),
        false => $this->supplierLinks[] = $link,
    };
}
```

**Reply by joshhanley** [reply]
@ju5t no worries! Ah interesting, I wonder why push works but not add 🤔

Yeah good to hear it's working 😁
